### PR TITLE
fix #59: when editing an exisiting project, add base starter if no dependencies are selected.

### LINKED ISSUE #59: Failed to build after removing security dependency
1. Generate a maven project , select  "Security" dependency.
2. Open pom.xml file, verify current value of dependencies node.
3. Right click on the editing area of the file, choose Edit starters.
4. In the QuickPick box, unselect "Security", and proceed following the prompt messages.
5. mvn package
image

### PR DESCRIPTION
**Pull request by Eskibear**
When generating a project from Spring Initializr, it always gives you the base starter if no dependencies are selected.

So when we remove all the dependencies during editing, it should add the base starter back.

### CODE REVIEW

File: src/DependencyManager.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export class DependencyManager {
     }
 
     public getSelectedDependencies(): IDependency[] {
-        return this.selectedIds.map((id: string) => this.dict[id]);
+        return this.selectedIds.map((id: string) => this.dict[id]).filter(Boolean);
```
Comment (Eskibear) [inline src/DependencyManager.ts]: for robustness

File: src/Metadata.ts [hunk 2]
```diff
@@ -26,7 +27,7 @@ export namespace dependencies {
             const rawJSONString: string = await Utils.downloadFile(`${Utils.settings.getServiceUrl()}dependencies?bootVersion=${bootVersion}`, true, { Accept: "application/vnd.initializr.v2.1+json" });
             starters[bootVersion] = JSON.parse(rawJSONString);
         }
-        return starters[bootVersion];
+        return _.cloneDeep(starters[bootVersion]);
```
Comment (Eskibear) [inline src/Metadata.ts]: always return a copy, in case the caller modify the cached content.

File: src/Routines.ts [hunk 3]
```diff
@@ -186,8 +186,8 @@ export module Routines {
                 vscode.window.showInformationMessage("No changes.");
                 return;
             }
-            const msgRemove: string = (toRemove && toRemove.length) ? `Removing: [${toRemove.map(d => manager.dict[d].name).join(", ")}].` : "";
-            const msgAdd: string = (toAdd && toAdd.length) ? `Adding: [${toAdd.map(d => manager.dict[d].name).join(", ")}].` : "";
+            const msgRemove: string = (toRemove && toRemove.length) ? `Removing: [${toRemove.map(d => manager.dict[d] && manager.dict[d].name).filter(Boolean).join(", ")}].` : "";
```
Comment (Eskibear) [inline src/Routines.ts]: avoid potential NPE

File: src/Routines.ts [hunk 4]
```diff
@@ -196,6 +196,14 @@ export module Routines {
                 TelemetryHelper.finishStep(stepProceed);
             }
 
+            // add spring-boot-starter if no selected starters
```
Comment (Eskibear) [inline src/Routines.ts]: add default starter, same issue with https://github.com/spring-projects/sts4/issues/52

**Comment by snicoll** [comment]
The title can be edited to mention that this is about editing an existing project. start.spring.io will always give you the base starter if no dependencies are selected.

**Comment by Eskibear** [comment]
@snicoll changed as the suggestion.

**Pull request merged**